File: lib/ts/src/internal/generation/GenerateRandomMockStub.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { GenerateContext } from './GenerateContext.js';
import { TType } from '../types/TType.js';
import { generateRandomStruct } from './GenerateRandomStruct.js';
import { TUnion } from '../types/TUnion.js';

export function generateRandomMockStub(types: { [key: string]: TType }, ctx: GenerateContext): object {
    const functions: string[] = Object.keys(types)
        .filter((key) => key.startsWith('fn.') && !key.endsWith('.->'))
        .filter((key) => !key.endsWith('_'))
        .map((key) => key);

    functions.sort((fn1, fn2) => fn1.localeCompare(fn2));

    const index = ctx.randomGenerator.nextIntWithCeiling(functions.length);

    const selectedFn = functions[index];

    const selectedArgs = types[selectedFn] as TUnion;
    const selectedResult = types[selectedFn + '.->'] as TUnion;

    const argFields = selectedArgs.tags[selectedFn].fields;
    const okFields = selectedResult.tags['Ok_'].fields;

    const arg = generateRandomStruct(null, false, argFields, ctx.copy({ alwaysIncludeRequiredFields: false }));
    const okResult = generateRandomStruct(null, false, okFields, ctx.copy({ alwaysIncludeRequiredFields: false }));

    return {
        [selectedFn]: arg,
        '->': {
            Ok_: okResult,
        },
    };
}

File: lib/go/internal/types/GenerateRandomMockStub.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package types

import (
	"sort"
	"strings"
)

// GenerateRandomMockStub creates a pseudo-random mock stub based on Telepact schema types.
func GenerateRandomMockStub(typesMap map[string]TType, ctx *GenerateContext) any {
	if ctx == nil || ctx.RandomGenerator == nil {
		return nil
	}

	functions := make([]string, 0)
	for key := range typesMap {
		if strings.HasPrefix(key, "fn.") && !strings.HasSuffix(key, ".->") && !strings.HasSuffix(key, "_") {
			functions = append(functions, key)
		}
	}

	sort.Strings(functions)
	if len(functions) == 0 {
		return nil
	}

	index := ctx.RandomGenerator.NextIntWithCeiling(len(functions))
	selectedFnName := functions[index]

	selectedFn, ok := typesMap[selectedFnName].(*TUnion)
	if !ok {
		return nil
	}

	selectedResult, ok := typesMap[selectedFnName+".->"].(*TUnion)
	if !ok {
		return nil
	}

	argFields := selectedFn.Tags[selectedFnName].Fields
	okFields := selectedResult.Tags["Ok_"].Fields

	alwaysIncludeRequired := false
	arg := GenerateRandomStruct(nil, false, argFields, ctx.Copy(nil, nil, &alwaysIncludeRequired, nil, nil))
	okResult := GenerateRandomStruct(nil, false, okFields, ctx.Copy(nil, nil, &alwaysIncludeRequired, nil, nil))

	return map[string]any{
		selectedFnName: arg,
		"->": map[string]any{
			"Ok_": okResult,
		},
	}
}
